CONTRACTS: Add __CPROVER_frees clauses in the front end

remi-delmas-3000 opened the pull request:
This is extracted from the dynamic frames PR #6887 and contains the front-end modifications only.

Add the following to the front-end:
  - add `__CPROVER_frees` clause for function and loop contracts
    in the parser and internal data structures
  - add `__CPROVER_freeable` built-in function
  - add `__CPROVER_is_freeable` built-in predicate
  - add `__CPROVER_is_freed` built-in predicate

Effective support for the new clause type and related predicates will come in the next PR.
  
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+# Assigns Clauses
 
-## In Function Contracts
+An _assigns_ clause lets the user to specify which memory locations may be
```
s/to //

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+assigned to by a function or a loop.
 
-### Syntax
+The set of locations described by the contract is the union of the locations
```
Is a "location" here a "memory location?" Also, is "contract" well defined here? There is no prior mention of this in this particular document. Also, now that I re-read this paragraph: in what way does a contract describe a set of (memory) locations? Why does that even matter?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
 
-### Syntax
+The set of locations described by the contract is the union of the locations
+specified by each of its assigns clauses, or the empty set of no _assigns_
```
s/of /if /

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
-### Syntax
+The set of locations described by the contract is the union of the locations
+specified by each of its assigns clauses, or the empty set of no _assigns_
+clause is specified.
```
Previously, wasn't there some code that would automatically determine what to havoc, and, therefore, in a way, the memory locations being written to?

remi-delmas-3000 replied:
There is indeed an inference pass for loop contracts in case there isn't a user-specified clause in the loop contract. 
I'll make the distinction. 
We don't have inference for function contracts. Since we ultimately want to have function contracts as standalone entities we wouldn't know what instructions to infer the assigned memory locations from.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+specified by each of its assigns clauses, or the empty set of no _assigns_
+clause is specified.
+
+The assigns clause may only refer to locations that exist in the calling context
```
If a function is called in exactly two places, and both of them have a procedure-local variable `x` in scope, can I use `&x` in the assigns clause?

remi-delmas-3000 replied:
syntactically the targets must be expressions of function parameters or globals, but they can surely resolve to that procedure local `&x` at some call-site. However I'm not sure what it would mean to modify the address of a procedure local variable. Modifying the value stored at this address, yes.

tautschnig replied:
Apologies, I should have used `x` instead of `&x`. I don't see where the documentation makes explicit that such expressions must be function parameters or globals?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,144 @@
+```
+targets ::= conditional-target-group (';' conditional-target-group)* ';'?
+conditional-target-group ::= (condition ':')? target (',' target)*
+target ::= l-value
````
Is "l-value" a thing? I'm only aware of "lvalue."

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
 
-The following functions can be used in assigns clause to specify ranges of 
-assignable addresses.
+An l-value target `expr` with a complete type `expr_t` specifies that the range
```
See above: I think it's "lvalue," not "l-value." Also, this is C/C++ terminology, perhaps we can find a way to describe this in a less language-specific way?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
 [CPROVER Manual TOC](../../)
 
-# Assigns Clause
+# Assigns Clauses
```
Having read to the end of this section, I really think this could do with an initial paragraph that sets the scene and explains why this even is useful or necessary.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,144 @@
-specifies that all bytes starting from the given pointer and until the end of 
-the object are assignable:
+Similarly, given an l-value expression `expr` with a complete type `expr_t`,
+ `__CPROVER_typed_target(expr)` specifies that the range
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+__CPROVER_assignable_t __CPROVER_typed_target(expr_t expr);
````
So why do we need `__CPROVER_typed_target` at all? It seems that one could just directly use `expr` instead?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr, size)` 
-specifies that `size` bytes starting from the given pointer and until the end of the object are assignable.
-The `size` value must such that `size <= __CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr)` holds:
+To specify assignable memory locations
+that are interpreted as pointers by the program, one must use
+`__CPROVER_assigns(ptr)` or `__CPROVER_assigns(__CPROVER_typed_target(ptr))`.
```
So how is this different from the above, at least for the case where `ptr` is an lvalue?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+}
+
+void step(struct circular_buffer_t *buf)
+// correct
+__CPROVER_assigns(__CPROVER_typed_target(buf->current))
+// not correct
+__CPROVER_assigns(__CPROVER_object_upto(&(buf->current), sizeof(buf->current))
```
At this point in the document: 1) I have no idea what the correctness criterion is, i.e., in what sense is the second example _not_ correct? 2) `__CPROVER_object_upto` has not been introduced.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,144 @@
+__CPROVER_assignable_t __CPROVER_object_whole(void *ptr);
+```
+REMARK: this also includes bytes in the object that are before the address given
+by `ptr`;
````
Nit pick: why the semicolon at the end?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,144 @@
+__CPROVER_assignable_t __CPROVER_object_from(void *ptr);
+```
 
+```c
+const size_t SIZE = 100;
+
+void init_slice_all(char *arr)
+__CPROVER_assigns(__CPROVER_object_from(arr))
+{
+  for(size_t i=0; i<SIZE; i++)
+    arr[i] = 0;
+}
+```
````
I'm not sure what to take away from this example?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
+specifies that the range of `size` bytes of `o` starting at `ptr` are assignable:
+The `size` value must such that the range does not exceed the object boundary,
+that is, `__CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr) >= size` must hold:
```
Could we perhaps describe this as "`(char *)ptr + size` also points into `o`" to avoid the use of built-ins?

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,144 @@
+```
+### Function parameters
+
+The memory locations storing function parameters values are considered local
````
s/parameters/parameter/

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,144 @@
+### Function parameters
+
+The memory locations storing function parameters values are considered local
+to the function and are hence always assignable.
```
Hmm, yes, but the same is true of any other live object. You'll have to phrase this in terms of the constraints that a contract introduces.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -59,17 +153,18 @@ int sum(const uint32_t a, const uint32_t b, uint32_t* out)
 ```
 
-### Semantics
+## Semantics
````
Now this is really useful. Perhaps this needs to be put near the beginning of the document? As is, a lot of syntax is discussed, but the reader has no idea what all this means. See also my note on function parameters above: the statement in the document makes a lot more sense once the semantics have been described.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -119,30 +217,40 @@ int sum(const uint32_t a, const uint32_t b, uint32_t* out)
+```
+Where `R[args/params]`, `A[args/params]`, `E[args/params]` denote the contract
+clause expressions rewritten by substituting
+function parameters with the argyments passed at the call site.
````
s/argyments/arguments/

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,21 @@ void ansi_c_internal_additions(std::string &code)
+    // frame specifications for contracts
+    // Type that describes assignable memory locations
+    "typedef void " CPROVER_PREFIX "assignable_t;\n"
+    // Declares a range of bytes as assignable (internal representation)
```
What does "internal representation" mean here?

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,21 @@ void ansi_c_internal_additions(std::string &code)
+    "typedef void " CPROVER_PREFIX "assignable_t;\n"
+    // Declares a range of bytes as assignable (internal representation)
+    CPROVER_PREFIX "assignable_t " CPROVER_PREFIX "assignable(void *ptr, "
+    CPROVER_PREFIX "size_t size,"
+    CPROVER_PREFIX "bool is_ptr_to_ptr);\n"
```
Some extra indent would help here.

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,21 @@ void ansi_c_internal_additions(std::string &code)
+    "\n"
+    // frame specifications for contracts
+    // Type that describes assignable memory locations
+    "typedef void " CPROVER_PREFIX "assignable_t;\n"
```
Why is this typedef required, couldn't we just use `void` in all the places that use it?

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,21 @@ void ansi_c_internal_additions(std::string &code)
+    CPROVER_PREFIX "bool is_ptr_to_ptr);\n"
+    // Declares a range of bytes as assignable
+    CPROVER_PREFIX "assignable_t " CPROVER_PREFIX "object_upto(void *ptr, "
+    CPROVER_PREFIX "size_t size);\n"
```
Indent

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,21 @@ void ansi_c_internal_additions(std::string &code)
+    CPROVER_PREFIX "size_t size);\n"
+    // Declares bytes from ptr to the end of the object as assignable
+    CPROVER_PREFIX "assignable_t " CPROVER_PREFIX "object_from(void *ptr);\n"
+    // Declares the whole object pointer to by ptr
```
I think this comment needs some fixes.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -768,14 +768,37 @@ void c_typecheck_baset::typecheck_declaration(
-          ID_loop_entry,
-          CPROVER_PREFIX "loop_entry is not allowed in preconditions.");
-      };
+      auto check_history_expr =
+        [&](const exprt expr, std::string &clause_type) {
+          disallow_subexpr_by_id(
+            expr,
+            ID_old,
+            CPROVER_PREFIX "old is not allowed in " + clause_type + ".");
+          disallow_subexpr_by_id(
+            expr,
+            ID_loop_entry,
+            CPROVER_PREFIX "loop_entry is not allowed in " + clause_type + ".");
+        };
+
+      auto check_return_value =
+        [&](const exprt &expr, std::string &clause_type) {
+          const irep_idt id = CPROVER_PREFIX "return_value";
+
+          auto pred = [&](const exprt &expr) {
+            if(!can_cast_expr<symbol_exprt>(expr))
+              return false;
+
+            return to_symbol_expr(expr).get_identifier() == id;
+          };
+
+          if(!has_subexpr(expr, pred))
+            return;
+
+          error().source_location = expr.source_location();
+          error() << id2string(id) + " is not allowed in " << clause_type << "."
+                  << eom;
+          throw 0;
+        };
```
How about using (static) functions instead of those lambdas? The enclosing function is way too long already, and those lambdas make matters worse.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -768,14 +768,37 @@ void c_typecheck_baset::typecheck_declaration(
+          if(!has_subexpr(expr, pred))
+            return;
+
+          error().source_location = expr.source_location();
+          error() << id2string(id) + " is not allowed in " << clause_type << "."
+                  << eom;
+          throw 0;
```
With #5837 merged it is now preferable to throw `invalid_source_file_exception` in such a case (see that PR for examples). This is to avoid introducing (and eventually getting rid of) `throw 0`.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -822,7 +845,9 @@ void c_typecheck_baset::typecheck_declaration(
           typecheck_spec_function_pointer_obeys_contract(expr);
-          check_history_expr(expr);
+          std::string clause_type = "function pointer preconditions";
+          check_history_expr(expr, clause_type);
+          check_return_value(expr, clause_type);
```
Do we always do both of these? Or does one always entail the other? If so, this ought to be refactored into a single call.

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -847,7 +847,26 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
 }
 
-void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+void c_typecheck_baset::throw_on_side_effects(const exprt &expr)
```
Could this perhaps be made a `static` function instead of adding to the API?

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -847,7 +847,26 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+{
+  if(has_subexpr(expr, ID_side_effect))
+  {
+    error().source_location = expr.source_location();
+    error() << "side-effects not allowed in assigns clause targets"
+            << messaget::eom;
+    throw 0;
```
See above: please use `invalid_source_file_exceptiont`.

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -903,39 +922,91 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+  {
+    if(!can_cast_expr<conditional_target_group_exprt>(target))
+    {
+      must_throw = true;
```
I'm not sure this `must_throw` is even needed: I _believe_ (but would be happy to be proved wrong) that the type checker eventually fails when any output to `error` was produced.

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -903,39 +922,91 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
 
+void c_typecheck_baset::typecheck_conditional_targets(
+  exprt::operandst &targets,
+  const std::function<void(exprt &)> typecheck_target,
```
What's the point of this function pointer? It seems to have exactly one target function.

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -903,39 +922,91 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
 
   // fatal errors
-  if(target.type().id() == ID_empty)
+  bool is_empty_type = target.type().id() == ID_empty;
+  bool is_assignable_typedef =
+    target.type().get(ID_C_typedef) == CPROVER_PREFIX "assignable_t";
+  // only allow void type if it is the typedef CPROVER_PREFIX "assignable_t"
+  if(target.type().id() == ID_empty && !is_assignable_typedef)
   {
     error().source_location = target.source_location();
     error() << "void-typed expressions not allowed as assigns clause targets"
             << eom;
     throw 0;
   }
```
What if we actually allowed `void`-typed expressions? Would there really be anything wrong in doing that? Wouldn't that allow us to get rid of `__CPROVER_assignable_t`?

remi-delmas-3000 replied:
The type is meant to documents the fact that the function specifies sets of assignable locations.

But mostly it is used to distinguish functions that specify assigns clauses from other void functions and process them differently when dynamic frames are involved.

tautschnig replied:
I am not convinced that a `typedef` is the best way to address this problem. The number of permitted functions is very small, can't we just specifically check for those in each place that requires checking?

remi-delmas-3000 replied:
Today we only provide 4 built-ins returning `__CPROVER_assignable_t`,
but the dynamic frames PRs builds on top of this to allow users to write their own functions returning `__CPROVER_assignable_t`. These user-defined functions can call other functions returning `__CPROVER_assignable_t` (built-in or user-defined), so the set is not necessarily small.

I intended  `__CPROVER_assignable_t` as an opaque type that represents a set of memory locations. It is defined as void in the front-end, and gets mapped to an actual representation by the instrumentation pass.

nwetzler replied:
My two cents, but I like the `__CPROVER_assignable_t` approach for representing sets of memory locations.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1943,8 +1943,54 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
       // This is an undeclared function.
+
+      // Is it the polymorphic typed_target function ?
+      if(identifier == CPROVER_PREFIX "typed_target")
+      {
+        if(expr.arguments().size() != 1)
```
The remainder of the code in this function sets out a poor example, but: could the body of this `if` be factored out into a function of its own?

remi-delmas-3000 replied:
thanks, will do.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1943,8 +1943,54 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+          error().source_location = f_op.source_location();
+          error() << "expected 1 argument for " << CPROVER_PREFIX "typed_target"
+                  << " found " << expr.arguments().size() << eom;
+          throw 0;
```
See above: `invalid_source_file_exceptiont`

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1943,8 +1943,54 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+        typecheck_expr(arg0);
+        if(!is_assignable(arg0) || !arg0.get_bool(ID_C_lvalue))
+        {
+          error().source_location = arg0.source_location();
+          error() << "argument of " << CPROVER_PREFIX "typed_target"
+                  << "must be assignable" << eom;
+          throw 0;
```
See above: `invalid_source_file_exceptiont`

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1943,8 +1943,54 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+        const auto &size = size_of_expr(arg0.type(), *this);
+        if(!size.has_value())
+        {
+          error().source_location = arg0.source_location();
+          error() << "sizeof not defined for argument of "
+                  << CPROVER_PREFIX "typed_target"
+                  << " of type " << to_string(arg0.type()) << eom;
+          throw 0;
```
See above: `invalid_source_file_exceptiont`

nwetzler commented on regression/contracts/frees-clause-and-predicates-fail/test.desc:
```diff
@@ -0,0 +1,10 @@
+^SIGNAL=0$
+--
+--
+This test checks that the front end rejects __CPROVER_is_freed in preconditions.
```
Maybe add that the intention of the code is to resize an array.

remi-delmas-3000 replied:
done

nwetzler commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2200,6 +2249,32 @@ exprt c_typecheck_baset::do_special_functions(
+    if(expr.arguments().size() != 1)
+    {
+      std::ostringstream error_message;
+      error_message << CPROVER_PREFIX "is_freeable expects one operand; "
```
Are these errors checked in regression?  I don't remember seeing them in the tests.

remi-delmas-3000 replied:
tests added

nwetzler commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,161 @@
+
+In a function contract:
+```c
+int foo(char *arr1, char *arr2, size_t size)
````
Are these examples included as regression tests?  If so, perhaps there is some way to link to them.

remi-delmas-3000 replied:
no these are just embedded in the doc.

feliperodri commented on regression/contracts/frees-clause-and-predicates/test.desc:
```diff
@@ -0,0 +1,17 @@
+CORE
```
If we know that this verification fails for lack of support in our back-end, shouldn't we mark this test as `KNOWNBUG` and describe the final expected behavior in the test description? Same as for all others test cases designed to test the implementation of frees clauses.

remi-delmas-3000 replied:
if we turn it to KNOWNBUG is it still going to be run as part of contracts regression  ?

feliperodri replied:
They will be skipped until you turn then back to `CORE` in the next PR, which implements their correct behavior. Not a blocker for now, but in the future marking as `KNOWNBUG` can help you to avoid updating all regression tests again if you implement a different behavior on another PR.

remi-delmas-3000 replied:
In that case I'd rather keep the test in CORE since it demonstrate that trying to prove a post that depends on the partially implemented `__CPROVER_is_freed` predicate is expected to fail.

feliperodri commented on regression/contracts/frees-clause-for-loop-pass/test.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0$
+--
+--
+This test checks that frees clauses are parsed and typechecked for for loops.
```
Are there no warnings or failures for no (current) support of `__CPROVER_frees` in our back-end?

remi-delmas-3000 replied:
I removed the clause from loops in the parser and the related tests

feliperodri commented on regression/contracts/frees-clause-while-loop-fail/test.desc:
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--apply-loop-contracts
+^main.c.* error: frees clause target must be a pointer-typed expression or a call to a function returning void$
```
Does the error message says the name and the type of the parameter an user tried to use with `__CPROVER_frees`? It might be really helpful for debugging.

feliperodri commented on regression/contracts/frees-clause-while-loop-pass/main.c:
```diff
@@ -0,0 +1,20 @@
+  while(i <= size)
+    // clang-format off
+  __CPROVER_assigns(i, arr[2], __CPROVER_POINTER_OBJECT(arr))
+  __CPROVER_frees(arr)
```
What if I call `__CPROVER_frees` with a function returning `void`, but with no pointers in it. For example,
```c
void foo()
{
  int x;
}
...
__CPROVER_frees(foo())
```

feliperodri commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -868,6 +890,14 @@ void c_typecheck_baset::typecheck_declaration(
           assigns.swap(lambda);
         }
 
+        typecheck_spec_frees(code_type.frees());
+        for(auto &frees : code_type.frees())
+        {
+          lambda_exprt lambda{temporary_parameter_symbols, frees};
+          lambda.add_source_location() = frees.source_location();
+          frees.swap(lambda);
+        }
```
Nitpick: we might want to throw a simple warning saying there is current no implementation for it.

feliperodri commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -845,6 +857,12 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+
+  if(code.find(ID_C_spec_frees).is_not_nil())
+  {
+    typecheck_spec_frees(
+      static_cast<unary_exprt &>(code.add(ID_C_spec_frees)).op().operands());
```
Why there is no coverage here?

feliperodri commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -1025,6 +1053,51 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
+
+    if(!can_cast_expr<symbol_exprt>(funcall.function()))
+    {
+      throw invalid_source_file_exceptiont(
+        "function pointer calls not allowed in frees clauses",
+        target.source_location());
+    }
```
We could add a regression test with a function pointer to cover this case.

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -492,7 +492,9 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
     // already type checked
   }
-  else if(expr.id() == ID_C_spec_assigns || expr.id() == ID_target_list)
+  else if(
+    expr.id() == ID_C_spec_assigns || expr.id() == ID_C_spec_frees ||
+    expr.id() == ID_target_list)
```
The lack of support here is also weird. Do we even need this if statement here?

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2268,6 +2270,30 @@ exprt c_typecheck_baset::do_special_functions(
+    {
+      error().source_location = f_op.source_location();
+      error() << CPROVER_PREFIX "is_freeable expects one operand; "
+              << expr.arguments().size() << "provided." << eom;
```
We could add a regression test to cover this case.

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2268,6 +2270,30 @@ exprt c_typecheck_baset::do_special_functions(
+      error().source_location = f_op.source_location();
+      error() << CPROVER_PREFIX "is_freeable expects one operand; "
+              << expr.arguments().size() << "provided." << eom;
+      throw 0;
```
Isn't there an exception specific for invalid parameters? `invalid_input_exceptiont`.

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2268,6 +2270,30 @@ exprt c_typecheck_baset::do_special_functions(
+  }
+  else if(identifier == CPROVER_PREFIX "is_freed")
+  {
+    if(expr.arguments().size() != 1)
+    {
+      error().source_location = f_op.source_location();
+      error() << CPROVER_PREFIX "is_freed expects one operand; "
+              << expr.arguments().size() << "provided." << eom;
+      throw 0;
```
Same as above.

feliperodri commented on src/ansi-c/parser.y:
```diff
@@ -2479,10 +2485,13 @@ iteration_statement:
+            static_cast<exprt &>(parser_stack($$).add(ID_C_spec_loop_invariant)).operands().swap(parser_stack($9).operands());
+
+          if(!parser_stack($10).operands().empty())
+            static_cast<exprt &>(parser_stack($$).add(ID_C_spec_decreases)).operands().swap(parser_stack($10).operands());
```
Why no coverage in these parser cases?

feliperodri commented on src/ansi-c/parser.y:
```diff
@@ -3396,6 +3410,27 @@ cprover_contract_assigns_opt:
+        {
+          $$=$1;
+          set($$, ID_C_spec_frees);
+          parser_stack($$).add_to_operands(exprt(ID_target_list));
```
Why no coverage?

feliperodri commented on src/util/rename_symbol.cpp:
```diff
@@ -171,6 +171,14 @@ bool rename_symbolt::rename(typet &dest) const
+      static_cast<const exprt &>(dest.find(ID_C_spec_frees));
+    if(spec_frees.is_not_nil() && have_to_rename(spec_frees))
+    {
+      rename(static_cast<exprt &>(dest.add(ID_C_spec_frees)));
+      result = false;
```
I think we don't have coverage here, because this will be use later in the back-end implementation. IS that accurate?

feliperodri commented on src/util/replace_symbol.cpp:
```diff
@@ -237,6 +237,11 @@ bool replace_symbolt::replace(typet &dest) const
 
+    const exprt &spec_frees =
+      static_cast<const exprt &>(dest.find(ID_C_spec_frees));
+    if(spec_frees.is_not_nil() && have_to_replace(spec_frees))
+      result &= replace(static_cast<exprt &>(dest.add(ID_C_spec_frees)));
```
I think we don't have coverage here, because this will be use later in the back-end implementation. IS that accurate?

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+A _frees clause_ allows the user to specify a set of pointers that may be freed
+by a function or the body of a loop. A function or loop contract contract may
+have zero or more frees clauses. When no clause is provided the empty set is
+used as default. When more than one frees clause is given, the sets of pointers
```
```suggestion
used as default (i.e., no memory will be freed). When more than one frees clause is given, the sets of pointers
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,147 @@
+                    | __CPROVER_freeable(lvalue-expr)
+```
+
+A frees clause target must be eiter:
````
```suggestion
A frees clause target must be either:
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+# Frees Clauses
+
+A _frees clause_ allows the user to specify a set of pointers that may be freed
+by a function or the body of a loop. A function or loop contract contract may
```
```suggestion
by a function or the body of a loop. A function or loop contract may
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+
+### Examples
+
+In a function contract
```
```suggestion
In a function contract:
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+pointers specified by the contract.
+
+### For replacement of function calls or loops by contracts
+When replacing a function call or a loop by a contract, each pointer of the
+_frees_ clause is non-deterministically freed after the function call
+or after the loop.
```
I think it is important to remember here that the ensures clauses could be used after to signal if a memory region will always be freed at the end of a function contract.

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+## Specifying parametric sets of freeable pointers using C functions
+
+Users can define parametric sets of freeable pointers by writing functions that
+return the built-in type void and call the built-in function
```
```suggestion
return the built-in type `void` and call the built-in function
```

feliperodri replied:
+1

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,147 @@
+```
+adds the given pointer to the freeable set described by the enclosing function.
+
+Calls to functions returning void can then be used as targets in
````
```suggestion
Calls to functions returning `void` can then be used as targets in
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+
+## Frees clause related predicates
+
+The predicate:
```
No need for this line.

remi-delmas-3000 replied:
fixed

feliperodri replied:
```suggestion
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,147 @@
+of the `free` function from `stdlib.h`, that is if and only if the pointer
+points to a valid dynamically allocated object and has offset zero.
+
+The predicate:
```
No need for this line.

remi-delmas-3000 replied:
fixed

feliperodri replied:
```suggestion
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,147 @@
+```
+can only be used in pre and post conditions, in contract checking or replacement
+modes. It returns `true` if and only if the pointer satisfies the preconditions
+of the `free` function from `stdlib.h`, that is if and only if the pointer
````
We should add a link to the preconditions in our library.

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,147 @@
+```
+can only be used in pre and post conditions, in contract checking or replacement
+modes. It returns `true` if and only if the pointer satisfies the preconditions
+of the `free` function from `stdlib.h`, that is if and only if the pointer
````
```suggestion
of the `free` function from [stdlib.h](https://github.com/diffblue/cbmc/blob/af8e92462c5e0f09a1b870d16879e4a2dda075f1/src/ansi-c/library/stdlib.c#L269-290), i.e., if and only if the pointer
```

remi-delmas-3000 replied:
is this a permalink ?

feliperodri replied:
yes

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,120 @@
+A frees clause target must be eiter:
+- an lvalue expression with a pointer type,
+- a call to the built-in function `__CPROVER_freeable`
+- a call to a user-defined side effect free and deterministic function returning
```
```suggestion
- a call to an user-defined side effect free and deterministic function returning
```

tautschnig replied:
(no, that would be wrong)

Commit:
CONTRACTS: Add __CPROVER_frees clauses

Add the following to the front-end:
- add `__CPROVER_frees` clause for function contracts
- add `__CPROVER_freeable` built-in function
- add `__CPROVER_is_freeable` built-in predicate
- add `__CPROVER_was_freed` built-in predicate

Effective support for the new clause type and related predicates
will come in a later PR.

tautschnig commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,120 @@
+# Frees Clauses
+
+A _frees clause_ allows the user to specify a set of pointers that may be freed
+by a function. A function contract may have zero or more frees clauses.
```
Is that just this function that has the contract attached to it, or any function that may (transitively) be called by the function that has the contract?

remi-delmas-3000 replied:
transitively

tautschnig commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,120 @@
+                    | __CPROVER_freeable(lvalue-expr)
+```
+
+A frees clause target must be eiter:
````
s/eiter/either/

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,120 @@
+
+## Semantics
+
+The set of pointers specified by the frees clause of the contract is interpreted
+at the function call-site.
```
At each call site of the function?

remi-delmas-3000 replied:
at each call site where the contract is checked or replaced

tautschnig commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,120 @@
+When replacing a function call by a contract, each pointer of the
+_frees clause_ is non-deterministically freed after the function call.
+
+## Specifying parametric sets of freeable pointers using C functions
```
Are there any restrictions on what control flow is permitted in such functions? For example, are recursion or loops without constant bound permitted?

remi-delmas-3000 replied:
in this first version recursion is prohibited, the function body is going to be inlined, and any remaining loops must be unwound to completion before the analysis. Adding this to the documentation.

tautschnig commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,120 @@
+```c
+__CPROVER_bool __CPROVER_is_freed(void *ptr);
+```
+can only be used in post conditions and returns `true` if and only if the
+pointer was freed during the execution of the function under analysis.
````
Will this also work when the pointer is re-allocated after having been freed? As in:
```
free(p);
p = malloc(1);
```
with the call to `malloc` returning the same adddress that `p` had before?

remi-delmas-3000 replied:
I'll change the name of the predicate to `__CPROVER_was_freed`. 

In a situation like this both `__CPROVER_was_freed(p)` and `__CPROVER_r_ok(p, 1)` will hold in post.

The instrumentation keeps track of deallocation events by recording ids of deallocated objects in a map, that gets passed to the predicate as a ghost parameter for evaluation.

if we wanted instead to assert that a pointer does *not* point into a live object as a post we would need to track the lifetime and current size of all objects in the program separately (without nondeterminism as is currently done).

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2268,6 +2270,30 @@ exprt c_typecheck_baset::do_special_functions(
+        expr.source_location()};
+    }
+    typecheck_function_call_arguments(expr);
+    return nil_exprt();
```
What to make of this?

remi-delmas-3000 replied:
I'm using the same approach as the one used with is_fresh : return `nil_exprt()` so that the original expression stays in place. When `do_special_function` returns a non-nil expression it replaces the original expression:

```cpp
  exprt tmp=do_special_functions(expr);

  if(tmp.is_not_nil())
    expr.swap(tmp);
  else
    typecheck_function_call_arguments(expr);
```

remi-delmas-3000 replied:
removing this completely since it seems like declaring the functions in `cprover_builtin_headers.h` and `ansi_c_internal_additions.cpp` suffices to have them type-checked. I don't understand why this was done like this for `is_fresh`

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2268,6 +2270,30 @@ exprt c_typecheck_baset::do_special_functions(
+        expr.source_location()};
+    }
+    typecheck_function_call_arguments(expr);
+    return nil_exprt();
```
... and this?

remi-delmas-3000 replied:
same as above. I am just following the pattern set by someone else with `is_fresh`

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -3396,6 +3398,27 @@ cprover_contract_assigns_opt:
+          set($$, ID_C_spec_frees);
+          mto($$, $3);
+        }
+        | TOK_CPROVER_FREES '(' ')'
+        {
+          $$=$1;
+          set($$, ID_C_spec_frees);
+          parser_stack($$).add_to_operands(exprt(ID_target_list));
+        }
```
Is this case even necessary?

remi-delmas-3000 replied:
yes, this is meant to allows for empty frees clauses to be explicitly written in contracts instead of having to be forcibly omitted.

tautschnig commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -870,6 +870,13 @@ void goto_convertt::convert_loop_contracts(
     loop->condition_nonconst().add(ID_C_spec_assigns).swap(assigns.op());
   }
 
+  auto frees = static_cast<const unary_exprt &>(code.find(ID_C_spec_frees));
+  if(frees.is_not_nil())
+  {
+    PRECONDITION(loop->is_goto());
+    loop->condition_nonconst().add(ID_C_spec_frees).swap(frees.op());
+  }
```
The documentation suggested this was only permitted on function pre/post-conditions and not on loop invariants?

remi-delmas-3000 replied:
good catch, this is a leftover

tautschnig reviewed (approved):
Ok in principle, but I'm confused on some occasions. Please clarify.
